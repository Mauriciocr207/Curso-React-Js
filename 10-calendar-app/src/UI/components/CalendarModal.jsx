import {
  Modal,
  ModalContent,
  ModalHeader,
  ModalBody,
  ModalFooter,
} from "@nextui-org/modal";
import { Button } from "@nextui-org/button";
import { Card } from "@nextui-org/card";
import { Spinner } from "@nextui-org/spinner";
import { getNow, getToday } from "../helpers";
import {
  CalendarInput,
  CalendarTextArea,
  CalendarDatePicker,
} from "./calendarForm";
import { func, bool } from "prop-types";
import useForm from "../hooks/useForm";
import { calendarFormValidation } from "../validation";
import { useEffect, useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import {
  startDeleteEvent,
  startCreateEvent,
  startUpdateEvent,
} from "../../app/features/calendar";
import CalendarDeleteEventBtn from "./CalendarDeleteEventBtn";

const getInitialForm = (event) => {
  return (
    event || {
      title: "",
      notes: "",
      start: getNow(),
      end: getNow().add({ hours: 2 }),
    }
  );
};

export default function CalendarModal({ isOpen, onClose }) {
  const {
    current: event,
    isLoading,
    errorMessage,
  } = useSelector((state) => state.calendar);
  const { user } = useSelector((state) => state.auth);
  const dispatch = useDispatch();
  const [modifyPermission, setModifyPermission] = useState(true);

  const {
    title,
    notes,
    start,
    end,
    formState,
    setFormState,
    onInputChange,
    isFormValid,
    errors,
    onSubmitForm,
    onResetForm,
  } = useForm(getInitialForm(event), calendarFormValidation);

  useEffect(() => {
    if (!isLoading && !errorMessage) {
      onResetForm();
    }
  }, [isLoading]);

  useEffect(() => {
    setModifyPermission(!event || event?.user?.id === user.id);
    setFormState(getInitialForm(event));
  }, [event]);

  const handleSubmit = (e) => {
    e.preventDefault();
    onSubmitForm();
    if (isFormValid) {
      if (event) {
        dispatch(startUpdateEvent({ ...formState }));
      } else {
        dispatch(startCreateEvent({ ...formState }));
      }
    }
  };

  const handleDeleteBtnClick = () =>
    dispatch(startDeleteEvent({ ...formState }));

  const handleCloseModal = () => {
    onResetForm();
    onClose();
  };

  return (
    <Modal
      backdrop="blur"
      isOpen={isOpen}
      onClose={handleCloseModal}
      className="px-2"
      motionProps={{
        variants: {
          enter: {
            y: 0,
            opacity: 1,
            transition: {
              duration: 0.3,
              ease: "easeOut",
            },
          },
          exit: {
            y: -20,
            opacity: 0,
            transition: {
              duration: 0.2,
              ease: "easeIn",
            },
          },
        },
      }}
    >
      <ModalContent>
        {(onClose) => (
          <form data-testid="form" onSubmit={handleSubmit}>
            <ModalHeader className="flex justify-between items-center gap-1 text-2xl text-gray-800">
              {event ? (
                <>
                  {modifyPermission ? (
                    <>
                      Editar Evento
                      <CalendarDeleteEventBtn onClick={handleDeleteBtnClick} />
                    </>
                  ) : (
                    "Evento"
                  )}
                </>
              ) : (
                "Nuevo Evento"
              )}
            </ModalHeader>
            <ModalBody>
              {errorMessage && (
                <Card className="bg-danger text-white font-semibold px-5 py-2 text-center">
                  {errorMessage}
                </Card>
              )}
              <div className="flex flex-col gap-4">
                <h5 className="font-semibold text-gray-700">
                  {" "}
                  Selecciona tus fechas{" "}
                </h5>
                <CalendarDatePicker
                  label="start"
                  name="start"
                  value={start}
                  minValue={getToday()}
                  onChange={onInputChange}
                  isDisabled={!modifyPermission}
                />
                <CalendarDatePicker
                  label="end"
                  name="end"
                  value={end}
                  minValue={start}
                  onChange={onInputChange}
                  isInvalid={!!errors.end}
                  errorMessage={errors.end}
                  isDisabled={!modifyPermission}
                />
                <h5 className="font-semibold text-gray-700 mt-2 ">
                  {" "}
                  TÃ­tulo y notas{" "}
                </h5>
                <CalendarInput
                  label="title"
                  value={title}
                  onChange={onInputChange}
                  name="title"
                  isInvalid={!!errors.title}
                  errorMessage={errors.title}
                  isDisabled={!modifyPermission}
                />
                <CalendarTextArea
                  label="notes"
                  name="notes"
                  value={notes}
                  onChange={onInputChange}
                  isInvalid={!!errors.notes}
                  errorMessage={errors.notes}
                  isDisabled={!modifyPermission}
                />
              </div>
            </ModalBody>
            {modifyPermission && (
              <ModalFooter>
                <Button
                  color="danger"
                  variant="light"
                  onPress={onClose}
                  className="font-semibold"
                >
                  Cerrar
                </Button>
                <Button
                  data-testid="submit-btn"
                  disabled={isLoading}
                  type="submit"
                  className="bg-primary-500 text-white font-semibold"
                >
                  {isLoading ? (
                    <Spinner color="white" size="sm" />
                  ) : event ? (
                    "Guardar"
                  ) : (
                    "Crear"
                  )}
                </Button>
              </ModalFooter>
            )}
          </form>
        )}
      </ModalContent>
    </Modal>
  );
}

CalendarModal.propTypes = {
  isOpen: bool.isRequired,
  onClose: func.isRequired,
};
