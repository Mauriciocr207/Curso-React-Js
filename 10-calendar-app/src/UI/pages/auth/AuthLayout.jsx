import "notyf/notyf.min.css";

export default function AuthLayout({ children }) {
  return (
    <section className="bg-gray-100 dark:bg-gray-900">
      <div className="flex flex-col items-center justify-center px-6 py-8 mx-auto h-screen lg:py-0-w-[30rem]">
        { children }
      </div>
    </section>
  )
}