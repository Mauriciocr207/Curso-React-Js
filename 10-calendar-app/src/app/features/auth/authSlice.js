import { createSlice } from '@reduxjs/toolkit'

const initialState = {
  isLoading: true,
  isAuthenticated: false,
  user: null,
  errorMessage: null,
}

export const authSlice = createSlice({
  name: 'auth',
  initialState,
  reducers: {
    login: (state, action) => {
      state.user = action.payload;
      state.isLoading = false;
      state.isAuthenticated = true;
      state.errorMessage = null;
    },
    logout: (state) => {
      state.user = null;
      state.isAuthenticated = false;
      state.isLoading = false;
    },
    load: (state) => {
      state.isLoading = true
    },
    setErrorMessage: (state, action) => {
      state.errorMessage = action.payload;
    }
  }
});

export const {
  login,
  logout,
  load,
  setErrorMessage
} = authSlice.actions

export default authSlice.reducer