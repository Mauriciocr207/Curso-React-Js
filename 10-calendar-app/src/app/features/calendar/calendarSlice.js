import { createSlice } from "@reduxjs/toolkit";
import { getNow } from "../../../UI/helpers";

const initialState = {
  events: [
    {
      _id: new Date().getTime(),
      title: "CumpleaÃ±os del Jefe",
      notes: "Comprar pastel",
      start: getNow(),
      end: getNow().add({ hours: 2 }),
      bgColor: "#fafafa",
      user: {
        _id: "123",
        name: "Fernando",
      },
    },
  ],
  current: null,
  isModalOpen: false,
};

const calendarSlice = createSlice({
  name: "calendar",
  initialState,
  reducers: {
    openModal: (state) => {
      state.isModalOpen = true;
    },
    closeModal: (state) => {
      state.isModalOpen = false;
    },
    setCurrentEvent: (state, action) => {
      state.current = action.payload;
    },
    createEvent: (state, action) => {
      state.events.push({
        _id: new Date().getTime(),
        user: {
          _id: "123",
          name: "Fernando",
        },
        ...action.payload,
      });
      state.isModalOpen = false;
    },
    updateEvent: (state, action) => {
      const event = action.payload;
      const eventIndex = state.events.findIndex(({_id}) => _id === event._id);
      state.events[eventIndex] = event;
      state.isModalOpen = false;
    },
    deleteEvent: (state, action) => {
      const { _id } = action.payload;
      state.events = state.events.filter(event => event._id !== _id);
      state.isModalOpen = false;
    }
  },
});

export const {
  setCurrentEvent,
  openModal,
  closeModal,
  createEvent,
  updateEvent,
  deleteEvent
} = calendarSlice.actions;

export default calendarSlice.reducer;
