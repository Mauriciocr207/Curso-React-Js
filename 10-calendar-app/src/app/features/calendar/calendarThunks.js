import { parseZonedDateTime } from "@internationalized/date";
import calendarApi from "../../api/calendarApi";
import {
  closeModal,
  createEvent,
  deleteEvent,
  loading,
  openModal,
  setCurrentEvent,
  setErrorMessage,
  setInitialEvents,
  updateEvent,
} from "./calendarSlice";

export const openCreateEventModal = () => {
  return (dispatch) => {
    dispatch(setCurrentEvent(null));
    dispatch(openModal());
  };
};

export const openUpdateEventModal = (event) => {
  return (dispatch) => {
    dispatch(setCurrentEvent(event));
    dispatch(openModal());
  };
};

export const startGetEvents = () => {
  return async (dispatch) => {
    try {
      const { data } = await calendarApi.get("/events");

      const events = data.events.map((event) => {
        return {
          ...event,
          start: parseZonedDateTime(event.start),
          end: parseZonedDateTime(event.end),
        };
      });

      dispatch(setInitialEvents(events));
    } catch (error) {
      console.log(error);
    }
  };
};

export const startCreateEvent = (event) => {
  return async (dispatch, getState) => {
    try {
      dispatch(loading());

      const newEvent = {
        ...event,
        start: event.start?.toString(),
        end: event.end?.toString(),
      };

      const { data } = await calendarApi.post("/events", newEvent);

      const { eventSaved } = data;

      eventSaved.start = event.start;
      eventSaved.end = event.end;
      eventSaved.user = getState().auth.user;

      dispatch(createEvent(eventSaved));
      dispatch(closeModal());
    } catch (error) {
      if (error.name === "AxiosError") {
        const { errors } = error.response.data;

        if (errors) {
          if (errors.start) {
            dispatch(setErrorMessage(errors.start.msg));
          } else if (errors.end) {
            dispatch(setErrorMessage(errors.end.msg));
          } else if (errors.title) {
            dispatch(setErrorMessage(errors.title.msg));
          }
        }

        setTimeout(() => {
          dispatch(setErrorMessage(null));
        }, 2000);
      }
    }
  };
};

export const startUpdateEvent = (event) => {
  return async (dispatch) => {
    try {
      dispatch(loading());
      const { id } = event;

      const newEvent = {
        ...event,
        start: event.start?.toString(),
        end: event.end?.toString(),
      };

      const { data } = await calendarApi.put(`/events/${id}`, newEvent);

      const { event:eventSaved } = data;

      eventSaved.start = event.start;
      eventSaved.end = event.end;

      dispatch(updateEvent(eventSaved));
      dispatch(closeModal());
    } catch (error) {
      if (error.name === "AxiosError") {
        const { errors, error:responseErr } = error.response.data;

        if(responseErr) {
          dispatch(setErrorMessage(responseErr.message));
        }

        if (errors) {
          if (errors.start) {
            dispatch(setErrorMessage(errors.start.msg));
          } else if (errors.end) {
            dispatch(setErrorMessage(errors.end.msg));
          } else if (errors.title) {
            dispatch(setErrorMessage(errors.title.msg));
          }
        }

        setTimeout(() => {
          dispatch(setErrorMessage(null));
        }, 2000);
      }
    }
  };
};

export const startDeleteEvent = (event) => {
  return async(dispatch) => {
    try {
      dispatch(loading());
      const { id } = event;
      await calendarApi.delete(`/events/${id}`);
      dispatch(deleteEvent(event));
    } catch (error) {
      console.log(error);
    }
  }
}