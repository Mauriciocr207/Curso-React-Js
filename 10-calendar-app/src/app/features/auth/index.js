import authSlice from "./authSlice"

export * from "./authThunks";
export {
    authSlice
}