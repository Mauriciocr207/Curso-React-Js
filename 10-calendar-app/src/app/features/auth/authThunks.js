import calendarApi from "../../api/calendarApi";
import { removeAllEvents } from "../calendar/calendarSlice";
import { load, login, logout, setErrorMessage } from "./authSlice"

export function startLogin({ email, pass }) {
    return async (dispatch) => {
        try {
            dispatch(load())
            
            const response = await calendarApi.post('/auth/login', { email, pass });

            const user = response.data;

            localStorage.setItem('token', user.token);
            
            dispatch(login({ name: user.name, id: user.id }));
        } catch (error) {
            if(error.name === "AxiosError") {
                dispatch(setErrorMessage("Credenciales invÃ¡lidas"));
                setTimeout(() => {
                    dispatch(setErrorMessage(null));
                }, 2000);
            }

            dispatch(logout());
        }
    }
}

export function startRegister({ name, email, pass }) {
    return async (dispatch) => {
        try {
            dispatch(load());

            const { data: user } = await calendarApi.post('/auth/register', { name, email, pass });

            localStorage.setItem('token', user.token);
            
            dispatch(login({ name: user.name, id: user.id }));
        } catch (error) {
            if(error.name === "AxiosError") {
                const { error: responseErr } = error.response.data;
                if(responseErr) {
                    dispatch(setErrorMessage(responseErr.message));
                }

                const { errors } = error.response.data;
                
                if(errors) {
                    if(errors.email) {
                        dispatch(setErrorMessage(errors.email.msg));
                    } else if(errors.name) {
                        return dispatch(setErrorMessage(errors.name.msg));
                    } else if(errors.pass) {
                        dispatch(setErrorMessage(errors.pass.msg));
                    }
                }
                
                setTimeout(() => {
                    dispatch(setErrorMessage(null));
                }, 2000);
            }

            dispatch(logout());
        }
    }    
}

export function checkAuthToken() {
    return async(dispatch) => {
        try {
            dispatch(load());
            const token = localStorage.getItem('token');
            if(!token) {
                return dispatch(logout());
            }

            const { data: user  } = await calendarApi.get('/auth/re-validate-token');

            localStorage.setItem('token', user.token);

            dispatch(login({ name: user.name, id: user.id }));
        } catch (error) {
            console.log(error);
        }
    }
}

export function startLogout() {
    return (dispatch) => {
        localStorage.removeItem('token');
        dispatch(logout());
        dispatch(removeAllEvents());
    }
}