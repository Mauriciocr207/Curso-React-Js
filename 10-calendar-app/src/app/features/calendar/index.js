import calendarSlice from "./calendarSlice";

export *  from "./calendarThunks";
export * from "./calendarSlice";
export {
    calendarSlice
}