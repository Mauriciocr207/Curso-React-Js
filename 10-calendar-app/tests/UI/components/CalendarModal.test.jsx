import { fireEvent, render, screen } from "@testing-library/react";
import CalendarModal from "../../../src/UI/components/CalendarModal";
import { Provider } from "react-redux";
import calendarSlice, {
  setCurrentEvent,
} from "../../../src/app/features/calendar/calendarSlice";
import authSlice from "../../../src/app/features/auth/authSlice";
import { initialState } from "../../fixtures/calendar/calendarFixtures";
import { authenticatedState } from "../../fixtures/auth/authFixtures";
import { configureStore } from "@reduxjs/toolkit";
import { event } from "../../fixtures/calendar/eventFixtures";
import dotenv from "dotenv";
dotenv.config({ path: ".env.test" });

const mockStartCreateEvent = jest.fn();
const mockStartUpdateEvent = jest.fn();
const mockStartDeleteEvent = jest.fn();

jest.mock("../../../src/app/features/calendar/calendarThunks", () => ({
  startCreateEvent: (event) => () => mockStartCreateEvent(event),
  startUpdateEvent: (event) => () => mockStartUpdateEvent(event),
  startDeleteEvent: (event) => () => mockStartDeleteEvent(event),
}));

const mockStore = configureStore({
  reducer: { auth: authSlice, calendar: calendarSlice },
  preloadedState: {
    calendar: initialState,
    auth: authenticatedState,
  },
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware({
      serializableCheck: false,
    }),
});

describe("tests on CalendarModal", () => {
  beforeEach(() => jest.clearAllMocks());
  test("renders without errors", () => {
    const { asFragment } = render(
      <Provider store={mockStore}>
        <CalendarModal isOpen={true} onClose={() => {}} />
      </Provider>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test("should call startCreateEvent when submitting form", async () => {
    render(
      <Provider store={mockStore}>
        <CalendarModal isOpen={true} onClose={() => {}} />
      </Provider>
    );

    fireEvent.change(screen.getByLabelText("title"), {
      target: { value: event.title },
    });

    fireEvent.change(screen.getByLabelText("notes"), {
      target: { value: event.notes },
    });

    fireEvent.submit(screen.getByTestId("form"));

    expect(mockStartCreateEvent).toHaveBeenCalledWith(
      expect.objectContaining({
        title: event.title,
        notes: event.notes,
      })
    );
  });

  test("should call startUpdatingEvent when submitting form", async () => {
    const newEvent = {
      title: "evento modificado",
      notes: "notas modificadas",
    };

    mockStore.dispatch(setCurrentEvent(event));

    render(
      <Provider store={mockStore}>
        <CalendarModal isOpen={true} onClose={() => {}} />
      </Provider>
    );

    fireEvent.change(screen.getByLabelText("title"), {
      target: { value: newEvent.title },
    });

    fireEvent.change(screen.getByLabelText("notes"), {
      target: { value: newEvent.notes },
    });

    fireEvent.submit(screen.getByTestId("form"));

    expect(mockStartUpdateEvent).toHaveBeenCalledWith(
      expect.objectContaining({
        title: newEvent.title,
        notes: newEvent.notes,
      })
    );
  });
});
