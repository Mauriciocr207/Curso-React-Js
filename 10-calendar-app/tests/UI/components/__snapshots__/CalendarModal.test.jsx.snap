// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests on CalendarModal renders without errors 1`] = `<DocumentFragment />`;
